gap/SnAnUnknownDegree.gi

```diff
@@ -0,0 +1,345 @@
+    tryThreeCycleCandidate := function()
+        local
+            # integer, loop variable
+            a,
+            # elements, in G
+            r, tPower, tPowerOld, c;
+        # Steps 2 & 3: New involution
+        # Check if we either tried enough conjugates or constructed enough
+        # three cycle candidates for the current involution t.
+        # If this is the case, we need to construct the next involution
+        if nrTriedConjugates >= C or nrThreeCycleCandidates >= T then
+            r := RandomElm(ri,"simplesocle",true)!.el;
+            a := 0;
+            tPower := r ^ M;
+            # Invariant: tPower = (r ^ M) ^ (2 ^ a)
+            repeat
+                a := a + 1;
+                tPowerOld := tPower;
+                tPower := tPower ^ 2;
+            until a = logInt2N or isone(ri)(tPower);
+            if a = logInt2N then
+                return NeverApplicable;
+            fi;
+            t := tPowerOld;
+            nrInvolutions := nrInvolutions + 1;
+            nrTriedConjugates := 0;
+            nrThreeCycleCandidates := 0;
+        fi;
+        # Steps 4 & 5: new three cycle candidate
+        # Try to construct a three cycle candidate via a conjugate of t. See
+        # the comment above this function.
+        nrTriedConjugates := nrTriedConjugates + 1;
+        c := t ^ RandomElm(ri,"simplesocle",true)!.el;
+        if not isequal(ri)(t * c, c * t) then
+            nrThreeCycleCandidates := nrThreeCycleCandidates + 1;
+            return (t * c) ^ 2;
+        else
+            return fail;
+        fi;
+    end;
+    # construct the iterator
+    oneThreeCycleCandidate := function()
+        local candidate;
+        repeat
+            if nrInvolutions >= B
+                and (nrTriedConjugates >= C or nrThreeCycleCandidates >= T)
+            then
+                # We are done and were not able to recognize Sn or An.
+                return TemporaryFailure;
+            fi;
+            candidate := tryThreeCycleCandidate();
+            if candidate = NeverApplicable then
+                return NeverApplicable;
+            fi;
+        until candidate <> fail;
+        return candidate;
+    end;
+    return oneThreeCycleCandidate;
+end);
+
+# G: the group to recognize
+# c: possibly a 3-cycle
+# returns a list of group elements. If G is isomorphic to an alternating or
+# symmetric group and c is a 3-cycle, then this function returns a list of
+# bolstering elements with respect to c.
+BindGlobal("BolsteringElements",
+function(ri, c, eps, N)
+    local result, R, S, prebolsteringElms, i, r, cr, cr2;
+    result := [];
+    R := Int(Ceil(7 / 4 * Log2(Float(eps ^ -1))));
+    S := 7 * N * R;
+    prebolsteringElms := [];
+    i := 0;
+    # find pre-bolstering elements
+    while i <= S and Length(prebolsteringElms) <= R do
+        r := RandomElm(ri,"simplesocle",true)!.el;
+        # test whether r is pre-bolstering
+        cr := c ^ r;
+        cr2 := c ^ (r ^ 2);
```

Comment:
Perhaps better to write it like this (we avoid one squaring)
```suggestion
        cr2 := cr ^ r;
```

Reply:
Done